funsor/affine.py

```diff
@@ -0,0 +1,172 @@
+            inputs.update(coeff.inputs)
+            inputs[name] = coeff.output
+
+        output = const.output
+        super(Affine, self).__init__(inputs, output)
+        self.coeffs = OrderedDict(coeffs)
+        self.const = const
+
+    def eager_subs(self, subs):
+        raise NotImplementedError("TODO")
+
+
+@eager.register(Affine, (Number, Tensor), tuple)
+def eager_affine(const, coeffs):
+    if not coeffs:
+        return const
+    return None
+
+
+@eager.register(Binary, Op, Affine, (Number, Tensor))
+def eager_binary_affine(op, lhs, rhs):
+    if op is ops.add or op is ops.sub:
+        const = op(lhs.const, rhs)
+        return Affine(const, tuple(lhs.coeffs.items()))
+    if op is ops.mul or op is ops.div:
+        const = op(lhs.const, rhs)
+        coeffs = tuple((name, op(coeff, rhs)) for name, coeff in lhs.coeffs.items())
+        return Affine(const, coeffs)
+    return None
+
+
+@eager.register(Binary, Op, (Number, Tensor), Affine)
+def eager_binary_affine(op, lhs, rhs):
+    if op is ops.add:
+        const = lhs + rhs.const
+        return Affine(const, tuple(rhs.coeffs.items()))
+    elif op is ops.sub:
+        return lhs + -rhs
+    if op is ops.mul:
+        const = lhs * rhs.const
+        coeffs = tuple((name, lhs * coeff) for name, coeff in rhs.coeffs.items())
+        return Affine(const, coeffs)
+    return None
+
+
+@eager.register(Binary, Op, Affine, Affine)
+def eager_binary_affine_affine(op, lhs, rhs):
+    if op is ops.add:
+        const = lhs.const + rhs.const
+        coeffs = lhs.coeffs.copy()
+        for name, coeff in rhs.coeffs.items():
+            if name in coeffs:
+                coeffs[name] += coeff
+            else:
+                coeffs[name] = coeff
+        return Affine(const, tuple(coeffs.items()))
+
+    if op is ops.sub:
+        return lhs + -rhs
+
+    return None
+
+
+@eager.register(Binary, Op, Affine, Variable)
+def eager_binary_affine_variable(op, affine, other):
+    if op is ops.add:
+        const = affine.const
+        coeffs = affine.coeffs.copy()
+        if other.name in affine.inputs:
+            coeffs[other.name] += 1.
+        else:
+            coeffs[other.name] = Number(1.)
+        return Affine(const, tuple(coeffs.items()))
+
+    if op is ops.sub:
+        return affine + -other
+
+    return None
+
+
+@eager.register(Binary, Op, Variable, Affine)
+def eager_binary_variable_affine(op, other, affine):
+    if op is ops.add:
+        return affine + other
+
+    if op is ops.sub:
+        return other + -affine
+
+    return None
+
+
+@eager.register(Unary, NegOp, Affine)
+def eager_negate_affine(op, affine):
+    const = -affine.const
+    coeffs = affine.coeffs.copy()
+    for name, coeff in coeffs.items():
+        coeffs[name] = -coeff
+    return Affine(const, tuple(coeffs.items()))
+
+
+#########################################
+# patterns for creating new Affine terms
+#########################################
```

Comment:
Looks like a bunch of other tests got broken because these change the output type of formerly lazy operations

Reply:
I wonder if we should restrict `Affine` to the `moment_matching` interpretation? Either way we should fix those tests.